### ISSUE TITLE
Add/Remove/Overwrite individual keys in the NvChad Plugins Setup tables?

### ISSUE THREAD

**Issue by erwin** [NvChad/NvChad#620]
**Is your feature request related to a problem? Please describe.**

What do you guys think about allowing each key of the plugin setup table to be overridden/added/removed, instead of entirely replacing the setup table the way you currently document?

```
M.plugins = {
   default_plugin_config_replace = {
      lspconfig = "custom.lspconfig",
   },
}
```

For example, I wanted to change add a few `sources` to the `cmp.lua` config. Instead of a way to append or remove from the cmp.setup.sources map, I had to make a copy of `cmp.lua` into `lua/custom/plugins/cmp.lua` and duplicate the whole file and make the changes.

```
   sources = {
      { name = "nvim_lsp" },
      { name = "luasnip" },
      { name = "buffer" },
      { name = "nvim_lua" },
      { name = "path" },
      { name = "neorg" },
   },

```

Similarly for bufferline.lua I wanted to add ordinal tab numbers, and had duplicate the all 143 lines of the bufferline file just to add a `numbers` function.

```
      numbers = function(opts)
        return string.format(
          '%s.%s', opts.lower(opts.id), opts.lower(opts.ordinal))
      end,
```

**Describe the solution you'd like**

It's not a big deal. NvChad works fine as it is.
In some cases, like `lsp-config` I think it's really better to just overwrite the whole thing.

Just wondering there's a way PackerCompile might be used so that an expensive operation like merging each key of a user provided table with each of the `NvChad` provided setup tables could be done without sacrificing any of the NvChad config's speed.

**Describe alternatives you've considered**
Just a thought. It's surely already possible to get the same result. Just would like to do something more granular, as for NvChad users that should be easier to maintain over time.

**Comment by siduck** [comment]
PR's are welcome! it'll take a while to add this stuff sincemy college is going on ;(

**Comment by zbirenbaum** [comment]
I made something like this for feline the other day when I was completely rewriting it and wanted the ability to swap around components fast. I would PR it but I completely changed every component and don't have time to make a version for the base config rn. I may work on it and others in a week or two after finals, but college be rough this time of year for everyone I think. gl @siduck ace those finals bro

**Comment by siduck** [comment]
there are tons of plugins and adding tables for each and every plugin would populate chadrc and make it more lengthy un-necessarily, the user can just copy default plugin config and edit it rather than us adding more code in the config which is not needed , dont be that lazy!! :angry:

**Comment by zbirenbaum** [comment]
I actually got this working just now for telescope but there are a few challenges. So the way this could work is that this is exclusively for small changes to the defaults, whereas a full config file would be required for big changes. If the value of the default override var in chadrc is not a string utils.override ignores it, inside the plugin config itself, the value of the default config override var is obtained again, and iterates the table replacing the values in the default or creating new ones.

The setup "defaults" table is declared prior to being passed to setup, and the replacement happens after that but prior to the actual setup. A more intelligent way to do this would be in override so that we don't have to do a get operation for the value twice for each default config, but that is a little tricky with the way override is currently programmed to function.

for example:
in chadrc:
```
default_plugin_config_replace = {
	telescope = {
		prompt_prefix = " ? ",
         }
         signature="custom.plugins.overrides.cmp_configs.lspsignature_cmp",
},
```
in telescope config
```
local defaults = {
	vimgrep_arguments = {
		 "rg",
        ...
        }
   ...
}
local override = require("core.utils").load_config().plugins.default_plugin_config_replace["telescope"]

if override ~= nil and type(override) == "table" then
	for k, v in pairs(override) do
		defaults[k] = v
	end
end

telescope.setup {
	defaults = defaults,
}
```

It works perfectly fine but i'm not a huge fan of having to call `require("core.utils").load_config().plugins.default_plugin_config_replace["telescope"]` in two different places. To avoid that would require a few backend changes on how overrides work but it shouldn't be too hard. I'm willing to take a stab at it over the next few days if you want to assign this to me since its still in the NvChad 2.0 milestones.

I would just implement and PR it except @siduck closed it but then it was reopened by @Akianonymus so I'm not sure whether its still a real goal for 2.0.

**Comment by siduck** [comment]
@zbirenbaum no worries, I'll work on this myself

**Comment by siduck** [comment]
@zbirenbaum I dont think this would be needed 

![image](https://user-images.githubusercontent.com/59060246/149604188-ebe1ddef-7728-4630-9100-c056f4f1a188.png)

if the user makes any changes to his custom table then 

- default table + custoom table ( merge them )

**Comment by zbirenbaum** [comment]
> @zbirenbaum I dont think this would be needed 
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/59060246/149604188-ebe1ddef-7728-4630-9100-c056f4f1a188.png)
> 
> 
> 
> if the user makes any changes to his custom table then 
> 
> 
> 
> - default table + custoom table ( merge them )
> 
> 
> 
> ```
> 
> vim.tbl_deep_extend('force',defaults,custom)
> 
> ```
> 
> 
> 
> (note : custom is the user's changed table!)

Yeah that looks right to me. I forgot the syntax for deep extend and wanted to get it working quickly as a proof of concept before I worried too much.

**Comment by zbirenbaum** [comment]
Addressed in #735

**Comment by ankitpal1029** [comment]
Hey @erwin is it possible for you to share how you got the tabs in bufferline to have those ordinal numbers? i tried copying the entire bufferline config and putting it inside default_plugin_config_replace but it didn't lead to anything
